File: Day 059/Problem 692. Top K Frequent Words.cpp
class Solution {
public:
    vector<string> topKFrequent(vector<string>& words, int k) {
        // Min heap
        priority_queue<pair<int, string>, vector<pair<int, string>>, decltype([](const auto &a, const auto &b) {
            if (a.first == b.first) return a.second < b.second;
            return a.first > b.first;
        })> pq;


        unordered_map<string, int> mp;
        for (string word : words) {
            mp[word]++;
        }

        for (auto entry : mp) {
            pq.push({entry.second, entry.first});
            if (pq.size() > k) {
                pq.pop();
            }
        }

        vector<string> result;
        while (!pq.empty()) {
            result.push_back(pq.top().second);
            pq.pop();
        }

        reverse(result.begin(), result.end());

        return result;
    }
};

File: Day 091/Problem 692. Top K Frequent Words.java
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.PriorityQueue;

class Solution {
    public List<String> topKFrequent(String[] words, int k) {
        Map<String, Integer> countMap = new HashMap<>();
        for (String word : words) {
            countMap.put(word, countMap.getOrDefault(word, 0) + 1);
        }

        PriorityQueue<Map.Entry<String, Integer>> pq = new PriorityQueue<>(
            (a, b) -> {
                if (a.getValue().equals(b.getValue())) {
                    return b.getKey().compareTo(a.getKey()); // greater lex come first
                } else {
                    return a.getValue() - b.getValue(); // min heap
                }
            }
        );

        for (Map.Entry<String, Integer> entry : countMap.entrySet()) {
            pq.offer(entry);
            if (pq.size() > k) {
                pq.remove();
            }
        }

        List<String> result = new ArrayList<>();
        while (!pq.isEmpty()) {
            result.add(pq.poll().getKey());
        }

        Collections.reverse(result);
        return result;
    }
}